lib/services/amazonka-dynamodb/src/Amazonka/DynamoDB/Internal.hs

```diff
@@ -0,0 +1,139 @@
+-- item whose key is a data type and the value is the data itself. We
+-- provide an actual sum type to interact with these.
+--
+-- For more information, see
+-- <https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/HowItWorks.NamingRulesDataTypes.html#HowItWorks.DataTypes Data Types>
+-- in the /Amazon DynamoDB Developer Guide/.
+data AttributeValue
+  = -- | An attribute of type List. For example:
+    --
+    -- @\"L\": [ {\"S\": \"Cookies\"} , {\"S\": \"Coffee\"}, {\"N\", \"3.14159\"}]@
+    L [AttributeValue]
+  | -- | An attribute of type Number Set. For example:
+    --
+    -- @\"NS\": [\"42.2\", \"-19\", \"7.5\", \"3.14\"]@
+    --
+    -- Numbers are sent across the network to DynamoDB as strings, to maximize
+    -- compatibility across languages and libraries. However, DynamoDB treats
+    -- them as number type attributes for mathematical operations.
+    NS [Text]
+  | -- | An attribute of type Map. For example:
+    --
+    -- @\"M\": {\"Name\": {\"S\": \"Joe\"}, \"Age\": {\"N\": \"35\"}}@
+    M (HashMap Text AttributeValue)
+  | -- | An attribute of type Null. For example:
+    --
+    -- @\"NULL\": true@
+    NULL
+  | -- | An attribute of type Number. For example:
+    --
+    -- @\"N\": \"123.45\"@
+    --
+    -- Numbers are sent across the network to DynamoDB as strings, to maximize
+    -- compatibility across languages and libraries. However, DynamoDB treats
+    -- them as number type attributes for mathematical operations.
+    N Text
+  | -- | An attribute of type Binary Set. For example:
+    --
+    -- @\"BS\": [\"U3Vubnk=\", \"UmFpbnk=\", \"U25vd3k=\"]@
+    BS [Base64]
+  | -- | An attribute of type Binary. For example:
+    --
+    -- @\"B\": \"dGhpcyB0ZXh0IGlzIGJhc2U2NC1lbmNvZGVk\"@
+    B Base64
+  | -- | An attribute of type String Set. For example:
+    --
+    -- @\"SS\": [\"Giraffe\", \"Hippo\" ,\"Zebra\"]@
+    SS [Text]
+  | -- | An attribute of type String. For example:
+    --
+    -- @\"S\": \"Hello\"@
+    S Text
+  | -- | An attribute of type Boolean. For example:
+    --
+    -- @\"BOOL\": true@
+    BOOL Bool
+  deriving (Eq, Read, Show, Generic, Hashable, NFData)
+
+$(makePrisms ''AttributeValue)
+
+instance FromJSON AttributeValue where
+  parseJSON = withObject "AttributeValue" $ \o ->
+    case KeyMap.toList o of
+      [("L", v)] -> L <$> parseJSON v
+      [("NS", v)] -> NS <$> parseJSON v
+      [("M", v)] -> M <$> parseJSON v
+      [("NULL", _)] -> pure NULL
+      [("N", v)] -> N <$> parseJSON v
+      [("BS", v)] -> BS <$> parseJSON v
+      [("B", v)] -> B <$> parseJSON v
+      [("SS", v)] -> SS <$> parseJSON v
+      [("S", v)] -> S <$> parseJSON v
+      [("BOOL", v)] -> B <$> parseJSON v
+      [] -> fail "No keys"
+      _ -> fail $ "Multiple or unrecognized keys: " ++ show (KeyMap.keys o)
+
+instance ToJSON AttributeValue where
+  toJSON = object . pure . toKeyValue
+  toEncoding = pairs . toKeyValue
+
+toKeyValue :: KeyValue kv => AttributeValue -> kv
```

Comment:
Hopefully this will be specialised and/or inlined.

Reply:
Got any suggestions to turn that "hopefully" into a "definitely"? I suppose I could just repeat myself.

Reply:
INLINE should be enough

Reply:
Couldn't get it to inline even with `{-# SPECIALIZE INLINE #-}`, inlining by hand instead.